g3doc/Tutorials/Toy/Ch-2.md

```diff
@@ -23,139 +23,320 @@ addresses this issue by being designed for extensibility. As such, there are
+// can define our operations. This is simply an ODS construct and does not
+// generate a class definition for the Toy dialect.
+def Toy_Dialect : Dialect {
+  // The namespace of our dialect.
+  let name = "toy";
+
+  // The c++ namespace that the dialect definition resides in.
+  let cppNamespace = "toy";
+}
+
+// Base class for toy dialect operations. This operation inherits from the base
+// `Op` class in OpBase.td, and provides:
+//   * The parent dialect of the operation.
+//   * The mnemonic for the operation, or the name without the dialect prefix.
+//   * A list of traits for the operation.
+class Toy_Op<string mnemonic, list<OpTrait> traits = []> :
+    Op<Toy_Dialect, mnemonic, traits>;
+
+// We define a toy operation by inherting from our base 'Toy_Op' class above.
+// Here we provide the mnemonic and a list of traits for the operation. The
+// constant operation is marked as 'NoSideEffect' as it is a pure operation
+// and may be removed if dead.
+def ConstantOp : Toy_Op<"constant", [NoSideEffect]> {
+  // Provide a summary and description for this operation. This can be used to
+  // auto-generate documenatation of the operations within our dialect.
+  let summary = "constant operation";
+  let description = [{
+    Generic calls represent calls to a user defined function that needs to
+    be specialized for the shape of its arguments. The callee name is attached
+    as a symbol reference via an attribute. The arguments list must match the
+    arguments expected by the callee. For example:
+
+      %0 = "toy.constant"()
+         { value = dense<[[1.0, 2.0, 3.0], [4.0, 5.0, 6.0]]> : tensor<2x3xf64> }
+        : () -> tensor<2x3xf64>
+
+    This is only valid if a function named "my_func" exists and takes two
+    arguments.
+  }];
+
+  // The generic call operation takes a symbol reference attribute as the
+  // callee, and two operands for the call.
+  let arguments = (ins SymbolRefAttr:$callee, F64Tensor:$lhs, F64Tensor:$rhs);
+
+  // The generic call operation returns a single value of TensorType.
+  let results = (outs F64Tensor);
+
+  // Add custom build methods for the constant operation. These method populates
+  // the `state` that MLIR uses to create operations, i.e. these are used when
+  // using `builder.create<ConstantOp>(...)`.
+  let builders = [
+    // Build a constant with a given constant tensor value.
+    OpBuilder<"Builder *builder, OperationState *result, "
+              "DenseElementsAttr value", [{
+      // Forward to the auto-generated build method.
+      build(builder, result, value.getType(), value);
+    }]>,
+    // Build a constant with a given constant floating-point value.
+    OpBuilder<"Builder *builder, OperationState *result, double value", [{
+      // Forward to a static function in the .cpp file.
+      buildConstantOp(builder, result, value);
+    }]>
+  ];
+
+  // Invoke a static `verify` method to verify this constant operation.
+  let verifier = [{ return ::verify(*this); }];
```

Comment:
Again, this seems all too complex to me for newcomers / beginners

Reply:
+1